.github/workflows/build.yml

```diff
@@ -0,0 +1,77 @@
+name: Build LLVM
+
+on:
+  push:
+    branches:
+      - 'rustc/*'
+  pull_request:
+    branches:
+      - 'rustc/*'
+  workflow_call:
+    outputs:
+      cache-key:
+        value: ${{ jobs.build.outputs.cache-key }}
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - id: cache-key
+        run: echo "cache-key=llvm-$(git rev-parse --short HEAD)-1" >> "$GITHUB_ENV"
```

Comment:
I think we probably want to do something smarter now. Unlike the usage in the bpf-linker repo, we should *always* use the cache here. The cache key should be just an encoding of the arguments we pass to cmake.

Put differently: imagine that we are applying our changes to a new LLVM branch from rustc -- we want to make use of whatever cached build products are available to us.

Then we should *always* run the LLVM build and upload the result.

There are also some vestiges of the reusable workflow here (see workflow_call above), which we probably no longer need.

Lastly: how are we going to consume the build product generated here from:
- the bpf-linker automation
- a user's local machine
?

Reply:
We could probably put a link in bpf-linker's README and tell them how to set the env var to pick it up. And we should probably also build for mac :P

Also I don't think we need to build dynamic libs? Nothing uses them and their usage is discouraged

Reply:
Dynamic libs are actually disabled here - dylibs are enabled by the `-DBUILD_SHARED_LIBS` option, which is not used here. I would prefer to stick to static ones even though they are bigger - as Alessandro said, dylibs are discouraged and I don't think we will hit any storage limits with the static ones.

About consuming the artifacts, I was thinking about having a `cargo xtask build` command which would by default pull the newest artifact from Github (we could just look for a newest artifact from the newest `rustc/*` branch). Of course only on `feature/fix-di` branch for now. Once everything is ready an we are going to release bpf-linker with BTF support, I think the best would be releasing binaries with the newest LLVM artifact, statically linked - in CI, on every tag. WDYT?

Reply:
I must have hallucinated I swear I saw those options 😂 Yeah I don't have a strong opinion, personally I'd just download the tarball somewhere and then set LLVM_SYS_BLA to point to it, but up to you

Reply:
The equivalent script in bpf-linker enables dynamic libs because I did in fact observe running out of space. Remember that each feature permutation links another copy of LLVM, and again for test binaries. It's not unusual for this to reach several gigabytes.

Regarding building for Mac: no need to cross compile, just run this workflow a second time on a Mac using a matrix strategy?

Reply:
> Regarding building for Mac: no need to cross compile, just run this workflow a second time on a Mac using a matrix strategy?

That won't work, because Github actions has **Intel** Mac runners, not **M1**.

Reply:
> The equivalent script in bpf-linker enables dynamic libs because I did in fact observe running out of space. Remember that each feature permutation links another copy of LLVM, and again for test binaries. It's not unusual for this to reach several gigabytes.

To be precise - you were running out of space when building? Here it doesn't seem to happen. And I don't think that the size of artifacts at the end exceeds 2 GB?

Reply:
Good point about cross compilation.

> To be precise - you were running out of space when building? Here it doesn't seem to happen.

It doesn't happen here because you're not building bpf-linker.

> And I don't think that the size of artifacts at the end exceeds 2 GB?

Best to check. Also, as I said, it is the size of the bpf-linker artifacts that is *much* bigger without dynamic linking.

I spent a lot of time making this work, I strongly encourage you to test thoroughly, it's about to become much harder to iterate once these things span two repos.